### PR TITLE
Feature/refactor mapper loader

### PR DESCRIPTION
**Pull request by markmcdowall**
## Description

Addition of the Mapper::Loader module to replace the XrefLoader.pm module from the previous xref pipeline. The changes include more tests and documentation for all the new functions.

## Benefits

Separated out the major update() function into testable chunks so that changes can be tested more easily

## Possible Drawbacks

Much of the code is the same, so there are probably places where there can be further work done to refine the code or the flow. This was an exercise to separate out the functionality into separate units.

## Testing

There are now lots more tests

### CODE REVIEW

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 1]
```diff
@@ -2062,12 +2049,762 @@ sub _get_alt_allele_hashes {
...
+
+=head2 get_valid_source_id_to_external_db_id
+  Description: Create a hash of all the external db names and ids that have
+               associated xrefs
+  Return type: Hashref
+  Caller     : internal
+
+=cut
+
+sub get_valid_source_id_to_external_db_id {
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: You might consider this having a cache for subsequent requests, although I'm not sure if it's ever used more than once per Hive process. Perhaps it's better if it's simple.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 2]
```diff
@@ -2062,12 +2049,762 @@ sub _get_alt_allele_hashes {
...
+        name         => $name,
+        where_from   => $where_from,
+        release_info => $release_info,
+      }
+    }
+  }
+}
+
+
+=head2 get_insert_identity_xref
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Confusing method name

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 3]
```diff
@@ -2062,12 +2049,762 @@ sub _get_alt_allele_hashes {
...
+  Return type: Iterator - Hashref
+  Example    : my $xref_dependents = $ba_handle->get_insert_dependent_xref();
+               while( my $xref_dependents_ref = $xref_dependents->() ) {
+                 my %xref_dependent = %{ $xref_dependents_ref };
+                 print "$xref_dependent{'xref_id'} - $xref_dependent{'acc'}";
+               }
+
+=cut
+
+sub get_insert_dependent_xref {
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: As above, this reads poorly in isolation

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 4]
```diff
@@ -2062,12 +2049,762 @@ sub _get_alt_allele_hashes {
...
+  my $sth = $self->dbi->prepare( $sql );
+  $sth->execute( $status ) || confess 'Could not set dumped status';
+
+  return;
+} ## end sub insert_process_status
+
+
+=head2 get_unmapped_reason
+  Description: Get the names and descriptions of valid reasons why an xref could
+               could be unmapped. These shoudld be used to describe the xrefs
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Typo

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 5]
```diff
@@ -2062,12 +2049,762 @@ sub _get_alt_allele_hashes {
...
+        version => $version,
+        desc    => $desc,
+        info    => $info,
+        type    => $type,
+        dbname  => $dbname
+      }
+    }
+  }
+} ## end sub get_insert_other_xref
+
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: I think BaseAdaptor is too big at this stage.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 6]
```diff
@@ -0,0 +1,1518 @@
...
+package Bio::EnsEMBL::Xref::Mapper::Loader;
+use strict;
+use warnings;
+use Carp;
+use Data::Dumper;
+use Bio::EnsEMBL::Utils::Exception;
+
+use parent qw( Bio::EnsEMBL::Xref::Mapper );
+
+use DBI;
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: This can go now. Data::Dumper above also.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 7]
```diff
@@ -0,0 +1,1518 @@
...
+  $self->xref->insert_process_status( 'core_loaded' );
+
+  return;
+} ## end sub update
+
+
+=head2 unmapped_xrefs_from_xrefdb_to_coredb
+  Arg [1]    : integer - $xref_offset
+  Arg [2]    : integer - $object_xref_offset
+  Arg [3]    : hashref - $analyss_ids
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Typo

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 8]
```diff
@@ -0,0 +1,1518 @@
...
+  Arg [4]    : hashref - $reason_ids
+  Description: Wrapper for running the functions for importing mapped xrefs from
+               unmapped xrefs from the xref db into the core db
+  Return type: Hashref
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Returntype seems to be undef

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 9]
```diff
@@ -0,0 +1,1518 @@
...
+  return;
+}
+
+
+=head2 map_xrefs_from_xrefdb_to_coredb
+  Arg [1]    : integer - $xref_offset
+  Arg [2]    : integer - $object_xref_offset
+  Description: Wrapper for running the functions for importing mapped xrefs from
+               the xref db into the core db
+  Return type: Hashref
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: This also appears to be out of date

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 10]
```diff
@@ -0,0 +1,1518 @@
...
+        $self->get_single_analysis( 'xrefchecksum' )
+      );
+    }
+    elsif ( $xref_handle{'type'} eq 'DEPENDENT' ) {
+      push @xref_list, $self->load_dependent_xref(
+        $xref_handle{'source_id'}, $xref_handle{'type'},
+        $xref_offset, $ex_id, $object_xref_offset );
+    }
+    else {
+      print "PROBLEM:: what type is $xref_handle{'type'}\n";
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: More colons than a digesting cow

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 11]
```diff
@@ -0,0 +1,1518 @@
...
+        analysis_id        => $analysis_id,
+        external_db_id     => $name_to_external_db_id{ $direct_handle{'dbname'} },
+        identifier         => $direct_handle{'acc'},
+        unmapped_reason_id => $reason_id } );
+
+      push @xref_list, $xref_id;
+    }
+  }
+
+  return @xref_list;
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Careful here. Passing a list by value can be expensive with large numbers of values.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 12]
```diff
@@ -0,0 +1,1518 @@
...
+        external_db_id     => $name_to_external_db_id{ $dependent_handle{'dbname'} },
+        identifier         => $dependent_handle{'acc'},
+        unmapped_reason_id => $reason_id,
+        parent             => $dependent_handle{'parent'} } );
+
+      push @xref_list, $xref_id;
+    }
+  }
+
+  return @xref_list;
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: As above, beware. We pass references to avoid triggering a copy on all list elements

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 13]
```diff
@@ -0,0 +1,1518 @@
...
+
+=cut
+
+sub get_xref_external_dbs {
+
+  my $self = shift;
+
+  my %externalname_to_externalid;
+
+  my $sth = $self->core->dbc->prepare('SELECT db_name, external_db_id FROM external_db');
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: We have a perfectly good DBEntryAdaptor for this kind of thing

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 14]
```diff
@@ -0,0 +1,1518 @@
...
+        'UNMAPPED',
+        $xref_handle{'info'},
+        $self->core->dbc);
+
+      $last_xref = $xref_id;
+      push @xref_list, $xref_id;
+    }
+
+    if ( defined $xref_handle{'ensembl_id'} ) {
+      $analysis_id= %{ $analysis_ids }{$xref_handle{'ensembl_object_type'}};
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: I wouldn't normally use this idiom. I prefer $hashref->{$key}, and I would need convincing that %{ $hashref }->{$key} is equivalent.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 15]
```diff
@@ -0,0 +1,1518 @@
...
+
+  if(@{$analysis_ids}) {
+    $analysis_id = $analysis_ids->[0];
+  }
+  else {
+    if ( $self->verbose ) {
+      print "No analysis with logic_name $logic_name found, creating ...\n";
+    }
+    # TODO - other fields in analysis table
+    $self->core()->dbc()->sql_helper()->execute_update(
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: You can probably do this with the AnalysisAdaptor too.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 16]
```diff
@@ -0,0 +1,1518 @@
...
+    $sth->execute;
+    my $max_val;
+    $sth->bind_columns( \$max_val );
+    $sth->fetch;
+    $meta_container->store_key_value($table . '_offset', $max_val || 0);
+    $sth->finish();
+
+    $results{ $table } = $max_val || 0;
+  }
+  return %results;
```
Comment (nerdstrike) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Again, returning in list context. You're copying the hash entire. Good for preventing side effects, bad for memory efficiency.

**Review by nerdstrike — CHANGES_REQUESTED** [review]
Ok... the tests and factorisation are good. There are a smattering of typos and a few oddities caused by Python encroachment. A nice parting gift from Mark.

I have not checked the thoroughness of the testing / edge-case handling.

Whomever is on this next should look into shrinking BaseAdaptor, as it's getting ridiculous. Also, there are situations where the core API could be used for non-bulk operations.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm (hunk 1 above)
Comment (markmcdowall) [reply modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: There are a couple of times that this happens, it came about so that the functions could be run in isolation as these functions were going to be run in the `update()` function. We could have functions like include a test to see if it has been pre-generated already and store within the object. This could probably be done in future work?

Comment (nerdstrike) [reply modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Absolutely. Functional trumps performance tuning.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm (hunk 5 above)
Comment (markmcdowall) [reply modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: I would tend to agree. There are probably places where there could be consolidation of functions, but for this exercise I was trying to just break down the functionality. This might also have to be an exercise for a future coding sprint.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 17]
```diff
@@ -2061,13 +2046,763 @@ sub _get_alt_allele_hashes {
...
+        cigar_line       => $cigar_line,
+        score            => $score,
+        evalue           => $evalue
+      }
+    }
+  }
+} ## end sub get_insert_identity_xref
+
+
+=head2 get_insert_checksum_xref
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Another confusing method name.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 18]
```diff
@@ -2061,13 +2046,763 @@ sub _get_alt_allele_hashes {
...
+
+=cut
+
+sub mark_mapped_xrefs {
+  my ( $self, $xref_list, $status ) = @_;
+
+  my $sql = (<<"SQL");
+    UPDATE xref
+    SET dumped = ?
+    WHERE xref_id IN ( @{[join',', ('?') x @{$xref_list}]} )
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: The second part of this line is really confusing and hard to read.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 19]
```diff
@@ -2061,13 +2046,763 @@ sub _get_alt_allele_hashes {
...
+  Return type: Iterator - Hashref
+  Example    : my $xrefs = $ba_handle->get_insert_direct_xref_low_priority();
+               while( my $xref_ref = $xrefs->() ) {
+                 my %xref = %{ $xref_ref };
+                 print $xref{'acc'};
+               }
+
+=cut
+
+sub get_insert_direct_xref_low_priority {
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Another confusing method name.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 20]
```diff
@@ -2061,13 +2046,763 @@ sub _get_alt_allele_hashes {
...
+  Return type: Iterator - Hashref
+  Example    : my $xrefs = $ba_handle->get_insert_dependent_xref_low_priority();
+               while( my $xref_ref = $xrefs->() ) {
+                 my %xref = %{ $xref_ref };
+                 print $xref{'acc'};
+               }
+
+=cut
+
+sub get_insert_dependent_xref_low_priority {
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Yet another confiusing method name... Actually, let's just say *all* the names in this package which begin with _get_insert_ are confusing to me, just in case I do not flag all of them during the review.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 21]
```diff
@@ -0,0 +1,1518 @@
...
+  my %name_to_external_db_id = $self->get_xref_external_dbs();
+
+  # Get the cutoff values
+  my %failed_sources = %{ $self->xref->get_unmapped_reason() };
+
+  my %summary_failed = %{ $failed_sources{'summary'} };
+  my %desc_failed    = %{ $failed_sources{'desc'} };
+
+  my %reason_id;
+  foreach my $key (keys %desc_failed){
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: A space before the curly brace, perhaps?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 22]
```diff
@@ -0,0 +1,1518 @@
...
+  Description: This is the main wrapper function to handle the extraction of
+               xrefs from the xref db and loading them into the core db.
+  Return type: Hashref
+  Example    : $loader_handle->update()
+
+=cut
+
+sub update {
+  my ( $self ) = @_;
+  # remove xref, object_xref, identity_xref, depenedent_xref, go_xref, unmapped_object, (interpro???), external_synonym, projections.
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: What is this line for?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 23]
```diff
@@ -0,0 +1,1518 @@
...
+    if(!defined $failed_id ) {
+      $failed_id = $self->add_unmapped_reason(
+        $summary_failed{$key}, $desc_failed{$key} );
+    }
+    $reason_id{$key} = $failed_id;
+  }
+
+  # my $verbose  = $self->verbose;
+  # my $core_dbi = $self->core->dbc;
+  # my $xref_dbi = $self->xref->dbc;
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: No commented-out code, please. In my experience, commented-out code has a tendency to rot, not to mention the fact most code analysers do not scan it.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 24]
```diff
@@ -0,0 +1,1518 @@
...
+  # my $core_dbh = $core_dbi->db_handle();
+
+  ####################
+  # Get analysis id's
+  ####################
+
+  my %analysis_ids = $self->get_analysis();
+  my $checksum_analysis_id; #do not populate until we know we need this
+
+  print "xref offset is $xref_offset, object_xref offset is $object_xref_offset\n";
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Do we need this?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 25]
```diff
@@ -0,0 +1,1518 @@
...
+=cut
+
+sub load_unmapped_dependent_xref {
+  my ( $self, $xref_offset, $analysis_id, $reason_id ) = @_;
+
+  my @xref_list = ();
+  my %name_to_external_db_id = $self->get_xref_external_dbs();
+
+  @xref_list = ();
+  my $last_acc= 0;
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Could use a space before the =.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 26]
```diff
@@ -0,0 +1,1518 @@
...
+  my $dependent_xref_handle = $self->xref->get_insert_dependent_xref_low_priority();
+  while ( my $dependent_handle_ref = $dependent_xref_handle->() ) {
+    my %dependent_handle = %{ $dependent_handle_ref };
+    if( !defined $name_to_external_db_id{ $dependent_handle{'dbname'} } ){
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Space after _if_?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm (hunk 14 above)
Comment (mkszuba) [reply modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Seconded. At the very least, _$hashref->{key}_ reads more easily than _%{hashref}{key}_.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 27]
```diff
@@ -0,0 +1,1518 @@
...
+        ensembl_object_type => $xref_handle{'ensembl_object_type'}
+      } );
+    }
+    else{
+      my $ensembl_object_type = 'Translation';
+      if ( $xref_handle{'seq_type'} eq 'dna' ) {
+        $ensembl_object_type = 'Transcript';
+      }
+
+      $analysis_id = %{ $analysis_ids }{ $xref_handle{'ensembl_object_type'} };
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Ditto.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 28]
```diff
@@ -0,0 +1,1518 @@
...
+      $last_xref = $xref_id;
+    }
+
+    my $object_xref_id = $self->add_object_xref(
+      $object_xref_offset,
+      $identity_xref_handle{'object_xref_id'},
+      $identity_xref_handle{'ensembl_id'},
+      $identity_xref_handle{'ensembl_type'},
+      $xref_id + $xref_offset,
+      $identity_xref_handle{'analysis_ids'}{ $identity_xref_handle{'ensembl_type'} },
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: I wonder if this could be made a bit more readable.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 29]
```diff
@@ -0,0 +1,1518 @@
...
+=cut
+
+sub load_dependent_xref {
+  my ( $self, $source_id, $type, $xref_offset, $ex_id, $object_xref_offset ) = @_;
+
+  my @master_problems;
+  my $err_master_count=0;
+  # $dependent_sth->execute($source_id, $type);
+  # my ($xref_id, $acc, $label, $version, $desc, $info, $object_xref_id, $ensembl_id, $ensembl_type, $master_xref_id);
+  # $dependent_sth->bind_columns(\$xref_id, \$acc, \$label, \$version, \$desc, \$info, \$object_xref_id, \$ensembl_id, \$ensembl_type, \$master_xref_id);
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Commented-out code.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 30]
```diff
@@ -0,0 +1,1518 @@
...
+        $object_xref_offset,
+        $dependent_xref_handle{'object_xref_id'},
+        $dependent_xref_handle{'ensembl_id'},
+        $dependent_xref_handle{'ensembl_type'},
+        $xref_id + $xref_offset,
+        $analysis_ids{ $dependent_xref_handle{'ensembl_type'} },
+        $self->core->dbc
+      );
+
+      if ( defined $dependent_xref_handle{'master_xref_id'} ) { # need to sort this out for FlyBase since there are EMBL direct entries from the GFF and dependent xrefs from Uniprot
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: I would move the comment to a separate line or lines, this one is long enough even without it present.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 31]
```diff
@@ -0,0 +1,1518 @@
...
+        # $add_dependent_xref_sth->execute(($object_xref_id+$object_xref_offset), ($master_xref_id+$xref_offset), ($xref_id+$xref_offset) );
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Commented-out code.

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 32]
```diff
@@ -0,0 +1,1518 @@
...
+        push @master_problems, $dependent_xref_handle{'acc'};
+        $err_master_count++;
+      }
+    }
+
+    $last_xref = $xref_id;
+    $last_ensembl = $dependent_xref_handle{'ensembl_id'};
+  }
+  if ( @master_problems ){
+    print "WARNING:: for $ex_id $err_master_count problem master xrefs\nExamples are :-\t";
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: Is there a point in just printing this, with no attempt to draw the user's attention?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 33]
```diff
@@ -0,0 +1,1518 @@
...
+
+  my $insert_sql = (<<'SQL');
+    INSERT INTO xref (
+      xref_id, external_db_id, dbprimary_acc, display_label, version,
+      description, info_type, info_text)
+    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+SQL
+
+  my $select_sth = $dbc->prepare( $select_sql );
+  my $insert_sth = $dbc->prepare( $insert_sql );
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: I take it abstaining from using _prepare_cached()_ in this module is intentional?

File: modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm [hunk 34]
```diff
@@ -0,0 +1,1518 @@
...
+
+  my $sth = $self->core->dbc->prepare('SELECT db_name, external_db_id FROM external_db');
+  $sth->execute() or confess( $self->core->errstr() );
+  while ( my @row = $sth->fetchrow_array() ) {
+    my $external_db_name = $row[0];
+    my $external_db_id   = $row[1];
+    $externalname_to_externalid{$external_db_name} = $external_db_id;
+  }
+
+  return %externalname_to_externalid;
```
Comment (mkszuba) [inline modules/Bio/EnsEMBL/Xref/Mapper/Loader.pm]: As @nerdstrike has already pointed out elsewhere, there are issues with not passing complex data structures by reference. Moreover, there are many other functions in this module which already use references for that so it might pay to be consistent.

File: modules/t/baseadaptor.t [hunk 35]
```diff
@@ -19,6 +19,8 @@ See the NOTICE file distributed with this work for additional information
 use strict;
 use warnings;
 
+use Data::Dumper;
```
Comment (mkszuba) [inline modules/t/baseadaptor.t]: I would say that if we do not use this (or any other) module in a package, better not to import it.

File: modules/t/loader.t [hunk 36]
```diff
@@ -0,0 +1,603 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+=cut
+
+use strict;
+use warnings;
+
+use Data::Dumper;
```
Comment (mkszuba) [inline modules/t/loader.t]: I would say that if we do not use this (or any other) module in a package, better not to import it.

File: modules/t/loader_mapped_xrefs.t [hunk 37]
```diff
@@ -0,0 +1,281 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+=cut
+
+use strict;
+use warnings;
+
+use Data::Dumper;
```
Comment (mkszuba) [inline modules/t/loader_mapped_xrefs.t]: I would say that if we do not use this (or any other) module in a package, better not to import it.

File: modules/t/loader_unmapped_xrefs.t [hunk 38]
```diff
@@ -0,0 +1,300 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+=cut
+
+use strict;
+use warnings;
+
+use Data::Dumper;
```
Comment (mkszuba) [inline modules/t/loader_unmapped_xrefs.t]: I would say that if we do not use this (or any other) module in a package, better not to import it.

File: modules/t/loader_update.t [hunk 39]
```diff
@@ -0,0 +1,283 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+=cut
+
+use strict;
+use warnings;
+
+use Data::Dumper;
```
Comment (mkszuba) [inline modules/t/loader_update.t]: I would say that if we do not use this (or any other) module in a package, better not to import it.

**Review by mkszuba — CHANGES_REQUESTED** [review]
Could use a few more changes... Nothing particularly serious though.

File: modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm [hunk 40]
```diff
@@ -482,41 +479,41 @@ sub upload_xref_object_graphs {
...
+      'acc'          => $xref->{ACCESSION},
+      'version'      => $xref->{VERSION} // 0,
+      'label'        => $xref->{LABEL}   // $xref->{ACCESSION},
+      'desc'         => $xref->{DESCRIPTION},
+      'source_id'    => $xref->{SOURCE_ID},
+      'species_id'   => $xref->{SPECIES_ID},
+      'info_type'    => $xref->{INFO_TYPE},
+      'info_text'    => $xref->{INFO_TEXT},
+      'update_label' => 1,
+      'update_desc' => 1
```
[the remark this answers is not in the document]
Comment (wbazant) [reply modules/Bio/EnsEMBL/Xref/DBSQL/BaseAdaptor.pm]: Also: please add a comma at the end of the line 492. Then if you add another line between 492 and 493, it will show up clearly in the diff - this is as important as aligning the arrows and all-round good Perl style.

File: .gitignore [hunk 41]
```diff
@@ -36,3 +36,4 @@ MultiTestDB.conf
 bioperl*.zip
 /modules/t/CLEAN.t
 /misc-scripts/xref_mapping/sql/populate_metadata.sql
+local/
```
[the remark this answers is not in the document]
Comment (markmcdowall) [reply .gitignore]: This is a side effect of using carton for handling the dependencies. It creates this directory.

**Comment by s-mm** [comment]
@mkszuba @nerdstrike What do we do with this PR?

**Comment by mkszuba** [comment]
Hah, I have completely forgotten about. Seeing as we are not going to backport mapper code to _ensembl_, we might as well just drop this PR.
